### DISCUSSION TITLE
How to listen to partial state changes (i.e. `select`) without hooks?

### DISCUSSION THREAD

**Discussion by wujek-srujek** [rrousselGit/riverpod#564] [Q&A] [answered]
I'm using `Riverpod` with `StateNotifier` in a Flutter app without `hooks`. The state managed by the notifier contains a few 
fields, and in some cases I just want to react to changes to only a part of the state.

```dart

enum TimerStatus {
  running,
  paused,
  // Stopped preemptively by the user.
  stopped,
  // Stopped naturally, i.e. it run to the end.
  completed,
}

@immutable
class TimerState {
  final Duration remaining;
  final TimerStatus status;

  const TimerState(
    this.remaining,
    this.status,
  );
}
```

The notifier and provider look like this:

```dart
class TimerNotifier extends StateNotifier<TimerState> {
  // ...
}

final timerNotifierProvider = StateNotifierProvider.autoDispose
    .family<TimerNotifier, TimerState, Duration>(
  (ref, interval) {
    // ...
  },
);
```

I tried using the `select` method in build using the `Consumer` widget:

```dart
final timerStatus = watch(
   timerNotifierProvider(widget.interval).select((value) => value.status),
);
```
but I can't get the code to compile with:
```
The argument type 'ProviderListenable<TimerStatus>' can't be assigned to the parameter type 'ProviderBase<Object?, dynamic>'.
```

Is this supposed to work?

Anyways, I made it work using a second provider that does the selection itself:

```dart
final timerStatusProvider =
    Provider.autoDispose.family<TimerStatus, Duration>(
  (ref, interval) => ref.watch(timerNotifierProvider(interval)).status,
);
```

and it works, but seems like quite some boilerplate to need to reimplement what `select` does. Is what I'm doing the right thing?
If `select` is only meant for hooks, shouldn't it be part (like an extension function or something) of `hooks_riverpod` to avoid confusion, like in my case?

**Comment by davidmartos96** [comment, marked as answer]
What version of riverpod are you running? Riverpod 1.0.0-dev supports `select` without the need of hooks.

![image](https://user-images.githubusercontent.com/22084723/124253244-1e645a00-db28-11eb-9925-70c0a9a17b46.png)

**Reply by wujek-srujek** [reply]
I'm using 0.14.0+3, 'latest' on Pub. I'm not ready to use dev (unstable?) releases - unless you tell me that this version is anyway more stable than 0.14.0+3.
1. When is it planned to have a stable 1.0.0 release?
2. Apart from upgrading, is a dedicated provider, like in my example, the right thing to do?

**Reply by rrousselGit** [reply]
> When is it planned to have a stable 1.0.0 release?

As soon as it is ready

> Apart from upgrading, is a dedicated provider, like in my example, the right thing to do?


Yes

**Reply by wujek-srujek** [reply]
> As soon as it is ready

I think a misunderstanding is on the way. I didn't mean to be rude or hasten things up, I was just curious about the possible time frame. A week, month, year? But if you don't know, that's also an answer. I do understand it is a project you work in your free time.
